### PR TITLE
Feature/25396 validerings api

### PR DESCRIPTION
**Pull request by cmoesgaard**
https://redmine.magenta-aps.dk/issues/25396

- [x] Have you updated the documentation?
- [x] Have you performed a smoke test of the application?
- [x] Have you written tests for your changes?
- [x] Have you updated the release notes?

### CODE REVIEW

File: frontend/src/api/Validate.js [hunk 1]
```diff
@@ -23,11 +30,25 @@ export default {
...
+      .post('/cpr/', payload).then(result => {
         return true
+      }, err => {
+        return createErrorPayload(err)
       })
-      .catch(r => {
-        return false
+  },
+
+  orgUnit (orgUnit, validity) {
```
Comment (danchr) [inline frontend/src/api/Validate.js]: `orgUnitUuid`?

File: frontend/src/views/organisation/workflows/MoOrganisationUnitTerminate.vue [hunk 2]
```diff
@@ -101,7 +102,14 @@ export default {
 
     ...mapGetters({
       orgUnitDetails: STORE_KEY + '/GET_DETAILS'
-    })
+    }),
+
+    validity () {
+      return {
+        'from': this.terminate.validity.to,
```
Comment (danchr) [inline frontend/src/views/organisation/workflows/MoOrganisationUnitTerminate.vue]: Add a comment explaining that this is an instantaneous validation, and hence the intentional duplicate `to`s?

File: backend/mora/service/address_handler/phone.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def _get_mo_properties(self):
     @classmethod
     def validate_value(cls, value):
         """Phone number is 8 digits, optionally with country code"""
-        if not re.match(r'\d{8}|\+45\d{8}', value):
+        if not re.match(r'^\d{8}$|^\+45\d{8}$', value):
```
Comment (danchr) [inline backend/mora/service/address_handler/phone.py]: Erm, do we truly want to reject foreign phone numbers? (Probably not related to this issue, but still…)

(FWIW I think the City of Aarhus has an office in Brussels.)

File: frontend/src/locales/da/alerts.json [hunk 4]
```diff
@@ -36,7 +36,12 @@
     "V_TERMINATE_UNIT_WITH_CHILDREN_OR_ROLES": "Kan ikke afslutte en enhed med aktive underenheder og roller",
     "V_CHANGING_THE_PAST": "Kan ikke foretage ændringer før dags dato",
     "E_INVALID_INPUT": "Ugyldigt input",
-    "E_UNKNOWN": "Uspecificeret fejl"
+    "E_UNKNOWN": "Uspecificeret fejl",
```
Comment (danchr) [inline frontend/src/locales/da/alerts.json]: I think we should retain this one as the last one.

File: frontend/src/locales/da/alerts.json [hunk 5]
```diff
@@ -36,7 +36,12 @@
...
+    "V_INVALID_ADDRESS_PHONE": "Feltet skal indeholde otte cifre",
+    "V_INVALID_ADDRESS_EMAIL": "Feltet skal indeholde en email-adresse",
```
Comment (danchr) [inline frontend/src/locales/da/alerts.json]: Lose the hyphen; it's one word.

File: frontend/src/locales/da/alerts.json [hunk 6]
```diff
@@ -36,7 +36,12 @@
...
+    "V_INVALID_ADDRESS_EAN": "Feltet skal indeholde 13 cifre",
+    "V_INVALID_ADDRESS_PNUMBER": "Feltet skal indeholde 10 cifre",
+    "V_INVALID_ADDRESS_WWW": "Feltet skal indeholde en korrekt webadresse"
```
Comment (danchr) [inline frontend/src/locales/da/alerts.json]: `gyldig`?

File: frontend/src/validators/ActiveEngagements.js [hunk 7]
```diff
@@ -14,7 +15,8 @@ export default {
     return Validate.activeEngagements(employee, validity)
   },
 
-  getMessage (value, args, data) {
-    return data
+  getMessage (field, args, key) {
+    let messages = i18n.messages[i18n.locale]
+    return messages.alerts.error[key]
```
Comment (danchr) [inline frontend/src/validators/ActiveEngagements.js]: What if we send back an untranslated error?

File: frontend/src/validators/ActiveEngagements.js [hunk 8]
```diff
@@ -14,7 +15,8 @@ export default {
     return Validate.activeEngagements(employee, validity)
   },
 
-  getMessage (value, args, data) {
-    return data
+  getMessage (field, args, key) {
```
Comment (danchr) [inline frontend/src/validators/ActiveEngagements.js]: This method is duplicated elsewhere; could we consolidate them into e.g. `i18n.js`?

File: NEWS.rst [hunk 9]
```diff
@@ -5,6 +5,8 @@ New features
 ------------
 
 * #27687: The various org funk relations now accept a ``user_key`` parameter.
+* #25396: Added backend validation API, for validating individual fields from
+  the front-end.
```
Comment (danchr) [inline NEWS.rst]: Perhaps mention that the frontend also uses these?

**Comment by danchr** [comment]
Oh, one thing; I think we should consider moving the validation into its own package, so that we could have the API and logic grouped together and easily discoverable.

File: backend/mora/service/employee.py [hunk 10]
```diff
@@ -75,15 +74,11 @@ def prepare_create(self, req):
...
 
-        userids = c.bruger.fetch(
-            tilknyttedepersoner="urn:dk:cpr:person:{}".format(cpr),
-            tilhoerer=org_uuid
-        )
+        valid_to = util.POSITIVE_INFINITY
 
-        if userids and userid not in userids:
-            exceptions.ErrorCodes.V_EXISTING_CPR(cpr=cpr)
+        validator.does_employee_with_cpr_already_exist(
```
[the remark this answers is not in the document]
Comment (cmoesgaard) [reply backend/mora/service/employee.py]: It'll prevent a person from being created through the "Ny medarbejder" twice, but it basically just mimics what we're already capable of.

[the remark this answers is not in the document]
Comment (danchr) [reply backend/mora/service/employee.py]: > It'll prevent a person from being created through the "Ny medarbejder" twice, but it basically just mimics what we're already capable of.

Agreed — just wanted to add a note concerning it 🙂

File: backend/mora/service/validate.py [hunk 11]
```diff
@@ -28,6 +28,9 @@ def org_unit_validity():
 
     .. :quickref: Validate; Validate org unit
 
+    :statuscode 200: Validation succeeded.
+    :statuscode 4XX: Validation failed.
```
Comment (danchr) [inline backend/mora/service/validate.py]: Won't this trigger a warning?

File: backend/mora/service/address_handler/phone.py (hunk 3 above)
Comment (cmoesgaard) [reply backend/mora/service/address_handler/phone.py]: I'm just reimplementing the functionality we already have. We can add support for arbitrary country codes, but that seems out of scope of this.

File: backend/mora/service/address_handler/www.py [hunk 12]
```diff
@@ -5,9 +5,21 @@
...
 
 
 class WWWAddressHandler(base.AddressHandler):
     scope = 'WWW'
     prefix = 'urn:magenta.dk:www:'
+
+    @classmethod
+    def validate_value(cls, value):
+        """Ensure value is correct URL"""
+        if not validators.url(value):
```
[the remark this answers is not in the document]
Comment (cmoesgaard) [reply backend/mora/service/address_handler/www.py]: I guess we could, but I already have `validators` as a requirement for email validation, so might as well use it for this as well, right?

File: backend/mora/service/validate.py [hunk 13]
```diff
@@ -9,15 +9,19 @@
...
+from .. import util
+from .. import validator
 
 blueprint = flask.Blueprint('validate', __name__, static_url_path='',
-                            url_prefix='/validate')
+                            url_prefix='/service/validate')
 
 
 @blueprint.route('/org-unit/', methods=['POST'])
+@util.restrictargs()
```
[the remark this answers is not in the document]
Comment (cmoesgaard) [reply backend/mora/service/validate.py]: I'd rather just be explicit about it, like we are with every other endpoint.

File: backend/mora/service/validate.py [hunk 14]
```diff
@@ -52,6 +86,25 @@ def employee_validity():
 def check_cpr():
     """
     Verify that an employee with the given CPR no. does not already exist
```
[the remark this answers is not in the document]
Comment (cmoesgaard) [reply backend/mora/service/validate.py]: With how it currently is, I think it makes sense to just check across all time periods. We never inactivate the actual users.

File: backend/mora/service/validate.py [hunk 15]
```diff
@@ -99,8 +190,33 @@ def employee_existing_associations():
     return flask.jsonify(success=True)
 
 
-@blueprint.route('/parent-org-unit/', methods=['POST'])
-def parent_org_unit():
+@blueprint.route('/candidate-parent-org-unit/', methods=['POST'])
```
[the remark this answers is not in the document]
Comment (cmoesgaard) [reply backend/mora/service/validate.py]: I guess, but we might as well be explicit :smile:

File: backend/mora/service/address_handler/dar.py [hunk 16]
```diff
@@ -97,23 +100,28 @@ def _fetch_from_dar(addrid):
...
+
+                r.raise_for_status()
+
+                if addrobjs:
+                    # found, escape loop!
+                    break
+            # The request mocking library throws a pretty generic exception
+            # catch and rethrow as something we know how to manage
+            except Exception as e:
+                raise LookupError(e)
```
Comment (danchr) [inline backend/mora/service/address_handler/dar.py]: The arguments to this exception should be something slightly more helpful; either the arguments of the caller, or a generic stringification of the message. Passing an exception to an exception is… unusual…

File: backend/mora/service/address_handler/www.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (danchr) [reply backend/mora/service/address_handler/www.py]: Ah, didn't notice that use!

File: backend/mora/service/address_handler/dar.py (hunk 16 above)
Comment (cmoesgaard) [reply backend/mora/service/address_handler/dar.py]: It gives the exact same result as before, we just handle it earlier, so it seems like it works okay.

**Pull request merged**